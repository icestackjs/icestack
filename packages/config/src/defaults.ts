import { components as defaultComponents } from '@icestack/preset-default/components'
import { base as defaultBase } from '@icestack/preset-default/base'
import { utilities as defaultUtilities } from '@icestack/preset-default/utilities'
import { defaultVarPrefix } from '@icestack/shared/constants'
import { generateColors } from '@icestack/theme-algorithm'
import type { CodegenOptions, BaseOptions, ComponentsOptions, ComponentsValue } from '@icestack/types'

export function getDefaultBase(options?: CodegenOptions) {
  const { base: baseOptions, mode: globalMode } = options ?? {}
  const { themes } = baseOptions ?? {}
  const base: {
    themes: {
      light?: object
      dark?: object
    }
  } & Partial<BaseOptions> = {
    themes: {},
    themeSelectorTemplate: (theme: string) => {
      return `[data-mode="${theme}"]`
    },
    generateColors,
    mediaDarkTheme: false // 'dark'
  }
  if (globalMode !== 'none') {
    if (themes?.light !== false) {
      // @ts-ignore
      base.themes.light = defaultBase.themes?.light
    }

    if (themes?.dark !== false) {
      // @ts-ignore
      base.themes.dark = defaultBase.themes?.dark
    }
  }

  return base as Partial<BaseOptions>
}

export function injectSchema(defaultComponents: ComponentsOptions, options?: CodegenOptions) {
  const { components, mode: globalMode } = options ?? {}
  if (globalMode === 'none') {
    return Object.entries(defaultComponents).reduce<ComponentsOptions>((acc, [key, opts]) => {
      const innerPreset = components && components[key] && (components[key] as Partial<ComponentsValue<Record<string, any>>>).mode === 'preset'
      if (innerPreset) {
        acc[key] = opts
      }
      return acc
    }, {})
  }
  return Object.entries(defaultComponents).reduce<ComponentsOptions>((acc, [key, opts]) => {
    const innerNone = components && components[key] && (components[key] as Partial<ComponentsValue<Record<string, any>>>).mode === 'none'

    acc[key] = innerNone ? {} : opts
    return acc
  }, {})
}

export function getCodegenDefaults(options?: CodegenOptions): Omit<CodegenOptions, 'outdir'> {
  const { mode: globalMode } = options ?? {}
  const base = getDefaultBase(options)
  const components = injectSchema(defaultComponents, options)
  const utilities = globalMode === 'none' ? {} : defaultUtilities
  return {
    mode: 'preset',
    pick: {
      base: true,
      styled: true,
      utils: true
    },
    log: true,
    dryRun: false,
    base,
    components,
    utilities,
    postcss: {
      varPrefix: {
        varPrefix: defaultVarPrefix
      },
      atMedia: {},
      selector: {}
    },
    cva: {
      format: 'ts',
      importFrom: '@icestack/cva'
    }
  }
}

export { sharedExtraColors, sharedExtraVars } from '@icestack/preset-default/base'
