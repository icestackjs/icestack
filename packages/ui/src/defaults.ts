import type { Config } from 'tailwindcss'
import { presetPrimaryColors, generateColorVars, sharedExtraColors, sharedExtraVars } from './base/colors'
import type { CodegenOptions, ComponentsOptions, CodegenMode, BaseOptions, ComponentsValue } from './types'
import { defaultVarPrefix } from './constants'
import { schemaMap, names as componentNames, defaultSelectorMap } from '@/components'
export { defaultSelectorMap } from '@/components'

export function getDefaultBase(mode?: CodegenMode) {
  const base: {
    themes: {
      light?: object
      dark?: object
    }
    themeSelectorTemplate: (theme: string) => string
    mediaDarkTheme: string | boolean
  } = {
    themes: {},
    themeSelectorTemplate: (theme: string) => {
      return `[data-mode="${theme}"]`
    },
    mediaDarkTheme: false // 'dark'
  }

  if (mode !== 'none') {
    base.themes.light = {
      selector: ':root',
      extraColors: sharedExtraColors.light,
      extraVars: sharedExtraVars,
      types: {
        primary: generateColorVars('primary', presetPrimaryColors.blue),
        success: generateColorVars('success', presetPrimaryColors.green),
        warning: generateColorVars('warning', presetPrimaryColors.gold),
        error: generateColorVars('error', presetPrimaryColors.red),
        neutral: generateColorVars('neutral', presetPrimaryColors.grey)
      }
    }
    base.themes.dark = {
      selector: '[data-mode="dark"]',
      extraColors: sharedExtraColors.dark,
      extraVars: sharedExtraVars,
      types: {
        primary: generateColorVars('primary', presetPrimaryColors.blue, true),
        success: generateColorVars('success', presetPrimaryColors.green, true),
        warning: generateColorVars('warning', presetPrimaryColors.gold, true),
        error: generateColorVars('error', presetPrimaryColors.red, true),
        neutral: generateColorVars('neutral', presetPrimaryColors.grey, true)
      }
    }
  }

  return base as Partial<BaseOptions>
}

export function injectSchema(map: ComponentsOptions, components?: ComponentsOptions) {
  return Object.entries(map).reduce<ComponentsOptions>((acc, [key, opts]) => {
    const k = key as unknown as (typeof componentNames)[number]
    acc[k] =
      components && components[key] && (components[key] as Partial<ComponentsValue<Record<string, any>>>).mode === 'none'
        ? {
            ...opts
          }
        : {
            ...opts,
            schema: schemaMap[k]?.schema
          }

    return acc
  }, {})
}

export function createDefaultTailwindcssExtends(opts: { varPrefix?: string } = {}): Config['theme'] {
  const { varPrefix = defaultVarPrefix } = opts
  return {
    borderRadius: {
      box: `var(${varPrefix}rounded-box, 1rem)`,
      btn: `var(${varPrefix}rounded-btn, 0.5rem)`,
      badge: `var(${varPrefix}rounded-badge, 1.9rem)`
    },
    minHeight: {
      6: '1.5rem',
      8: '2rem',
      12: '3rem',
      16: '4rem'
    },
    lineHeight: {
      1: '0.25rem',
      2: '0.5rem'
    }
  }
}

export function getCodegenDefaults(options?: CodegenOptions): Omit<CodegenOptions, 'outdir'> {
  const base = getDefaultBase(options?.mode)
  const components = options?.mode === 'none' ? {} : injectSchema(defaultSelectorMap, options?.components)
  return {
    mode: 'preset',
    pick: {
      base: true,
      styled: true,
      utils: true
    },
    log: true,
    dryRun: false,
    base,
    components,
    postcss: {
      varPrefix: {
        varPrefix: defaultVarPrefix
      },
      atMedia: {},
      selector: {}
    }
  }
}

export { sharedExtraColors, sharedExtraVars, generate, generateColorVars, gray, makeRgbaValue, presetPrimaryColors } from './base/colors'
