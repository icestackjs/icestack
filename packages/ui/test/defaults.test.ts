import { defaultSelectorMap, injectSchema } from '@/defaults'

describe('defaults', () => {
  it('injectSchema', () => {
    expect(injectSchema(defaultSelectorMap)).toMatchSnapshot()
  })
})
