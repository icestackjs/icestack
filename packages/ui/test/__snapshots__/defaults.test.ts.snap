// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defaults > injectSchema 1`] = `
{
  "alert": {
    "schema": [Function],
    "selector": ".alert",
  },
  "avatar": {
    "schema": [Function],
    "selector": ".avatar",
  },
  "badge": {
    "schema": [Function],
    "selector": ".badge",
  },
  "bottom-navigation": {
    "schema": [Function],
    "selector": ".btm-nav",
  },
  "breadcrumbs": {
    "schema": [Function],
    "selector": ".breadcrumbs",
  },
  "button": {
    "schema": [Function],
    "selector": ".btn",
  },
  "card": {
    "schema": [Function],
    "selector": ".card",
  },
  "carousel": {
    "schema": [Function],
    "selector": ".carousel",
  },
  "chat": {
    "schema": [Function],
    "selector": ".chat",
  },
  "checkbox": {
    "schema": [Function],
    "selector": ".checkbox",
  },
  "collapse": {
    "schema": [Function],
    "selector": ".collapse",
  },
  "countdown": {
    "schema": [Function],
    "selector": ".countdown",
    "varPrefix": {
      "ignoreProp": [
        "--value",
      ],
      "ignoreValueCustomProperty": [
        "--value",
      ],
    },
  },
  "diff": {
    "schema": [Function],
    "selector": ".diff",
  },
  "divider": {
    "schema": [Function],
    "selector": ".divider",
  },
  "dropdown": {
    "schema": [Function],
    "selector": ".dropdown",
  },
  "file-input": {
    "schema": [Function],
    "selector": ".file-input",
  },
  "form": {
    "schema": [Function],
  },
  "indicator": {
    "schema": [Function],
    "selector": ".indicator",
  },
  "input": {
    "schema": [Function],
    "selector": ".input",
  },
  "join": {
    "schema": [Function],
    "selector": ".join",
  },
  "kbd": {
    "schema": [Function],
    "selector": ".kbd",
  },
  "link": {
    "schema": [Function],
    "selector": ".link",
  },
  "loading": {
    "schema": [Function],
    "selector": ".loading",
  },
  "mask": {
    "schema": [Function],
    "selector": ".mask",
  },
  "menu": {
    "schema": [Function],
    "selector": ".menu",
  },
  "navbar": {
    "schema": [Function],
    "selector": ".navbar",
  },
  "progress": {
    "schema": [Function],
    "selector": ".progress",
  },
  "radial-progress": {
    "schema": [Function],
    "selector": ".radial-progress",
    "varPrefix": {
      "ignoreProp": [
        "--size",
        "--thickness",
        "--value",
      ],
      "ignoreValueCustomProperty": [
        "--size",
        "--thickness",
        "--value",
      ],
    },
  },
  "radio": {
    "schema": [Function],
    "selector": ".radio",
  },
  "range": {
    "schema": [Function],
    "selector": ".range",
  },
  "rating": {
    "schema": undefined,
    "selector": ".rating",
  },
  "select": {
    "schema": [Function],
    "selector": ".select",
  },
  "skeleton": {
    "schema": [Function],
    "selector": ".skeleton",
  },
  "stack": {
    "schema": [Function],
    "selector": ".stack",
  },
  "stat": {
    "schema": [Function],
    "selector": ".stat",
  },
  "steps": {
    "schema": [Function],
    "selector": ".step",
  },
  "swap": {
    "schema": [Function],
    "selector": ".swap",
  },
  "tab": {
    "schema": [Function],
    "selector": ".tab",
  },
  "table": {
    "schema": [Function],
    "selector": ".table",
  },
  "textarea": {
    "schema": [Function],
    "selector": ".textarea",
  },
  "timeline": {
    "schema": [Function],
    "selector": ".timeline",
  },
  "toast": {
    "schema": [Function],
    "selector": ".toast",
  },
  "toggle": {
    "schema": [Function],
    "selector": ".toggle",
  },
  "tooltip": {
    "schema": [Function],
    "selector": ".tooltip",
    "varPrefix": {
      "ignoreProp": [
        "--tooltip-tail",
        "--tooltip-color",
        "--tooltip-text-color",
        "--tooltip-tail-offset",
      ],
      "ignoreValueCustomProperty": [
        "--tooltip-tail",
        "--tooltip-color",
        "--tooltip-text-color",
        "--tooltip-tail-offset",
      ],
    },
  },
}
`;
